tokio/src/local.rs

```diff
@@ -0,0 +1,374 @@
+        assert!(self.is_current());
+        unsafe {
+            (*self.tasks.get()).insert(task);
+        }
+    }
+
+    fn release(&self, _: UnsendTask<Self>) {
+        unreachable!("tasks should only be completed locally")
+    }
+
+    fn release_local(&self, task: &UnsendTask<Self>) {
+        assert!(self.is_current());
+        unsafe {
+            (*self.tasks.get()).remove(task);
+        }
+    }
+
+    fn schedule(&self, task: UnsendTask<Self>) {
+        assert!(self.is_current());
+        unsafe {
+            (*self.queue.get()).push_front(task);
+        }
+    }
+}
+
+impl Scheduler {
+    fn new() -> Self {
+        Self {
+            tasks: UnsafeCell::new(task::OwnedList::new()),
+            queue: UnsafeCell::new(VecDeque::with_capacity(64)),
+        }
+    }
+
+    fn with<F>(&self, f: impl FnOnce() -> F) -> F {
+        struct Entered<'a> {
+            current: &'a Cell<Option<NonNull<Scheduler>>>,
+        }
+
+        impl<'a> Drop for Entered<'a> {
+            fn drop(&mut self) {
+                self.current.set(None);
+            }
+        }
+
+        CURRENT_TASK_SET.with(|current| {
+            current.set(Some(NonNull::from(self)));
+            let _entered = Entered { current };
+            f()
+        })
+    }
+
+    fn is_current(&self) -> bool {
+        CURRENT_TASK_SET
+            .try_with(|current| {
+                current
+                    .get()
+                    .iter()
+                    .any(|current| ptr::eq(current.as_ptr(), self as *const _))
+            })
+            .unwrap_or(false)
+    }
+
+    fn next_task(&self) -> Option<UnsendTask<Self>> {
+        unsafe { (*self.queue.get()).pop_front() }
+    }
+
+    fn tick(&self) {
+        for _ in 0..MAX_TASKS_PER_TICK {
+            let task = match self.next_task() {
+                Some(task) => task,
+                None => return,
+            };
+
+            if let Some(task) = task.run(&mut || Some(self.into())) {
+                self.schedule(task);
+            }
+        }
+    }
+}
+
+unsafe impl Send for Scheduler {}
+unsafe impl Sync for Scheduler {}
+
+impl fmt::Debug for Scheduler {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Scheduler { .. }").finish()
+    }
+}
+impl Drop for Scheduler {
+    fn drop(&mut self) {
+        // Drain all local tasks
+        while let Some(task) = self.next_task() {
+            task.shutdown();
+        }
+
+        // Release owned tasks
+        unsafe {
+            (*self.tasks.get()).shutdown();
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::runtime;
+
+    #[test]
+    fn local_current_thread() {
+        let mut rt = runtime::Builder::new().current_thread().build().unwrap();
+        TaskSet::new().block_on(&mut rt, async {
+            spawn_local(async {}).await.unwrap();
+        });
+    }
+
+    #[test]
+    fn local_threadpool() {
```

Comment:
I'm not sure how this can work given that `blocking` has the ability to give away the `Worker` to a different thread. Wouldn't that also give away the `LocalFuture`, which shouldn't be possible since it's not `Send`?

Reply:
Oh, that's definitely a concern. I think we may need some special handling for blocking...

Reply:
@jonhoo just did some testing, and I don't think this is an issue, at least currently: the thread that calls `block_on` on a runtime cannot issue in-place blocking calls anyway; attempts to call `runtime::blocking::in_place` inside of a `block_on` will panic regardless. Since `block_on` is the only way to run a local task set anyway, in-place blocking won't be an issue.

I started with a test like this:
```rust
    #[test]
    fn local_threadpool_blocking_call() {
        // use crate::sync::oneshot;
        thread_local! {
            static ON_RT_THREAD: Cell<bool> = Cell::new(false);
        }

        ON_RT_THREAD.with(|cell| cell.set(true));

        let mut rt = runtime::Runtime::new().unwrap();
        TaskSet::new().block_on(&mut rt, async {
            assert!(ON_RT_THREAD.with(|cell| cell.get()));
            let join = spawn_local(async move {
                assert!(ON_RT_THREAD.with(|cell| cell.get()));
                crate::runtime::blocking::in_place(|| {});
                assert!(ON_RT_THREAD.with(|cell| cell.get()));
            });
            join.await.unwrap();
        });
    }
```
I expected this to fail with the second assertion that the local task was still running on the original thread. However, it actually panics with:
```

running 1 test
thread 'runtime::local::tests::local_threadpool_blocking_call' panicked at 'can only call blocking when on Tokio runtime', src/libcore/option.rs:1190:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
thread 'runtime::local::tests::local_threadpool_blocking_call' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(...)', src/libcore/result.rs:1165:5
test runtime::local::tests::local_threadpool_blocking_call ... FAILED
```

To confirm, I wrote a test like this:
```rust
    #[test]
    fn block_on_rt_thread() {
        let mut rt = runtime::Runtime::new().unwrap();
        rt.block_on(async {
            crate::runtime::blocking::in_place(|| {});
        });
    }
```
which fails with the same panic.

If we eventually want to allow worker threads (rather than the thread creating the runtime) to create and run their own local task sets through something other than `block_on`, however, we will definitely need to think about the interaction with blocking.

Reply:
Ah, interesting. I suspect we will want that pattern to work eventually, though I'm not really sure how yet. Maybe leave that first test in there, and then add a `#[should_panic]`?

Reply:
Plus a comment explaining why it is marked as `#[should_panic]`.

Reply:
@jonhoo I've added documentation and tests as you described!